File: platforms/apple/Sources/BotaAppleSDK/Host/EncryptedUploadV2Host.swift
import BotaDeviceSDKC
import Foundation

protocol EncryptedUploadV2Host: Sendable {
    func execute(_ effect: CoreEffect) async -> AsyncThrowingStream<CoreHostEventPayload, Error>
}

struct EncryptedUploadV2HostFailure: Error, Equatable, Sendable {
    let errorCode: UInt32
    let retryable: Bool
    let protocolStatus: UInt16?
    let detail: String?

    init(errorCode: UInt32, retryable: Bool, protocolStatus: UInt16? = nil, detail: String? = nil) {
        self.errorCode = errorCode
        self.retryable = retryable
        self.protocolStatus = protocolStatus
        self.detail = detail
    }
}

struct UnavailableEncryptedUploadV2Host: EncryptedUploadV2Host {
    func execute(_ effect: CoreEffect) async -> AsyncThrowingStream<CoreHostEventPayload, Error> {
        AsyncThrowingStream { continuation in
            continuation.finish(throwing: EncryptedUploadV2HostFailure(
                errorCode: UInt32(BOTA_DEVICE_SDK_V1_ERROR_FEATURE_UNAVAILABLE),
                retryable: false,
                detail: "encrypted upload v2 native host is not configured"
            ))
        }
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/host/EncryptedUploadV2Host.kt
package dev.bota.sdk.internal.host

import dev.bota.sdk.internal.core.CoreCancellationId
import dev.bota.sdk.internal.core.CoreEffect
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

internal fun interface EncryptedUploadV2Host {
    fun execute(effect: CoreEffect): Flow<CoreHostEventPayload>

    suspend fun confirmationAttemptedOrClaimCancellation(cancellationId: CoreCancellationId): Boolean = false

    suspend fun cancel(cancellationId: CoreCancellationId) = Unit
}

internal open class EncryptedUploadV2HostException(
    val errorCode: UInt,
    val retryable: Boolean,
    val protocolStatus: UShort? = null,
    message: String,
) : IllegalStateException(message)

internal class EncryptedUploadV2ConfirmationException(
    val writeSucceeded: Boolean,
    message: String,
    cause: Throwable,
) : EncryptedUploadV2HostException(19u, false, message = message) {
    init { addSuppressed(cause) }
}

internal class UnavailableEncryptedUploadV2Host : EncryptedUploadV2Host {
    override fun execute(effect: CoreEffect): Flow<CoreHostEventPayload> = flow {
        throw EncryptedUploadV2HostException(
            errorCode = 7u,
            retryable = false,
            message = "encrypted upload v2 native host is not configured",
        )
    }
}
